feat: Implementing close_position method

ndatta-nethermind opened the pull request.

ndatta-nethermind committed:
feat: Implementing close_position method

ndatta-nethermind committed:
feat: Implementing close_position method

ndatta-nethermind committed:
feat: Implementing close_position method

ndatta-nethermind committed:
feat: Implementing close_position method

ndatta-nethermind committed:
Ndatta/megavault apis (#385)

varex83 committed:
fix: update links to direct to the new docs (#384)

varex83 committed:
chore: update indexer types with latest protocol changes (#387)

This PR:
- Updates `dydx-proto` version.
- Introduces `#[serde(deny_unknown_fields)]` to not miss any fields from
return types.
- Adds several missed return types.
- Reorders some of the struct fields to align with the indexer
implementation.
- Has some minor fixes for new clippy version

tyleroooo committed:
fix: make sure cjs is emitted in build (#388)

varex83 committed:
feat: add zero payments flag to funding payments (#390)

varex83 committed:
feat: add strict serde (#392)

lukabuz committed:
fix: url mapping of getParentSubaccount (#395)

Reference to the PR this one is created in place of
[here](https://github.com/dydxprotocol/v4-clients/pull/389)

varex83 committed:
fix: close position

varex83 committed:
fix: close position

varex83 committed:
fix: close position

varex83 committed:
fix: close position

varex83 committed:
Merge pull request #9 from NethermindEth/bohdan/fix-close-position

fix: close position

ndatta-nethermind committed:
Updated v4-proto version to 8.2 and projet version to 1.1.5 (#397)

ndatta-nethermind committed:
feat: Added example

ndatta-nethermind committed:
feat: Added example

ndatta-nethermind committed:
feat: Added example

ndatta-nethermind committed:
feat: Added example

ndatta-nethermind committed:
fix: Reformat

ndatta-nethermind committed:
fix: Reformat

ndatta-nethermind committed:
fix: Reformat

ndatta-nethermind committed:
fix: Reformat

lukabuz committed:
feat: add proxy support to JS client (#399)

This pull request introduces support for proxy configurations in the
`v4-client-js` package. The changes include adding proxy support to both
REST and WebSocket clients, updating constructors, and introducing
utility functions to handle proxy agents.

### Proxy Support Enhancements:

* **Proxy configuration added to `IndexerConfig`:** The `IndexerConfig`
class now includes an optional `proxy` parameter of type
`AxiosProxyConfig`, allowing users to specify proxy settings. The
constructor was updated to initialize this property.
(`v4-client-js/src/clients/constants.ts`,
[v4-client-js/src/clients/constants.tsR247-R252](diffhunk://#diff-1a0931f9cb2a865b6d73148c440e7d29f6fe1f6991fb406a190f2e56082bcbd5R247-R252))
* **REST client proxy integration:** The `RestClient` class now uses an
`axios` instance configured with the optional `proxy` parameter. This
replaces the previous direct use of the `request` function for HTTP
requests. (`v4-client-js/src/clients/modules/rest.ts`,
[[1]](diffhunk://#diff-0a6ead65c843f918b4acedfc6e9b595005415d093160fbdb43bca1c867dd49ecR1-R28)
[[2]](diffhunk://#diff-0a6ead65c843f918b4acedfc6e9b595005415d093160fbdb43bca1c867dd49ecL32-R39)
* **WebSocket client proxy integration:** The `SocketClient` class now
supports proxy configurations by utilizing the `HttpsProxyAgent`
library. The `proxyAgent` is initialized using a new utility function,
`getProxyAgent`, and passed to the WebSocket connection.
(`v4-client-js/src/clients/socket-client.ts`,
[[1]](diffhunk://#diff-c4c36d3ced6c5f292f1d45c1ff474d57f921370435bd85caafbe72b6465d260cR1-R5)
[[2]](diffhunk://#diff-c4c36d3ced6c5f292f1d45c1ff474d57f921370435bd85caafbe72b6465d260cR42)
[[3]](diffhunk://#diff-c4c36d3ced6c5f292f1d45c1ff474d57f921370435bd85caafbe72b6465d260cR57-R67)
* **Utility function for proxy agent creation:** A new utility function,
`getProxyAgent`, was added to generate an `HttpsProxyAgent` instance
from an `AxiosProxyConfig` object. (`v4-client-js/src/lib/utils.ts`,
[[1]](diffhunk://#diff-63d05f995846a96324239cfaf78bdd5bfb2b1e5a32726bb236915b6ce379e201R1-R3)
[[2]](diffhunk://#diff-63d05f995846a96324239cfaf78bdd5bfb2b1e5a32726bb236915b6ce379e201R80-R84)

### Dependency Updates:

* **Added `https-proxy-agent` dependency:** The `https-proxy-agent`
library was added to the `package.json` file to enable proxy support for
WebSocket connections. (`v4-client-js/package.json`,
[v4-client-js/package.jsonR65](diffhunk://#diff-ec61c6dbaf924b282dbb0275f615f0fe5daa0eab5e3f720e5cedb87fa3caeffaR65))

gaonip commented on v4-client-py-v2/examples/close_position_example.py:
```diff
@@ -0,0 +1,30 @@
+    market = Market(
+        (await indexer.markets.get_perpetual_markets(MARKET_ID))["markets"][MARKET_ID]
+    )
+    response = await node_client.close_position(
```
wouldn't you need to open a position first as an example, and then target that specific position to close?

ndatta-nethermind replied:
Updated the examples.

ndatta-nethermind committed:
Updated close_position example

ndatta-nethermind committed:
Updated close_position example

ndatta-nethermind committed:
Updated close_position example

ndatta-nethermind committed:
Updated close_position example

ndatta-nethermind committed:
Linted the changes

ndatta-nethermind committed:
Linted the changes

ndatta-nethermind committed:
Linted the changes

ndatta-nethermind committed:
Linted the changes

northstar456 committed:
fix: fix `permissoned keys` example (#405)

gaonip commented on v4-client-py-v2/dydx_v4_client/node/client.py:
```diff
@@ -1217,3 +1222,64 @@ async def register_affiliate(
+            side=order_side,
+            size=order_size,
+            price=0,
+            reduce_only=False,
```
reduce only should be true on a close_position

gaonip commented on v4-client-py-v2/dydx_v4_client/node/client.py:
```diff
@@ -1217,3 +1222,64 @@ async def register_affiliate(
+            time_in_force=None,
+            side=order_side,
+            size=order_size,
+            price=0,
```
does this work with a long?  dont think that works

ndatta-nethermind replied:
`long` means? Data type or `long term` order?

gaonip replied:
try to do a LONG side order, it wont work when price is 0, that's why you need to test this works
you can check on mintscan/indexer, whether your order was actually closed

gaonip commented on v4-client-py-v2/tests/test_mutating_node_client.py:
```diff
@@ -251,3 +254,17 @@ async def test_register_affiliate(node_client, wallet, test_address, recipient):
+    response = await node_client.close_position(
+        wallet, test_address, 0, market, None, random.randint(0, 1000000000)
+    )
+    print(response)
```
the test has to check whether the size is 0, whether position really closed

ndatta-nethermind committed:
fix: Set reduce_only field to true

ndatta-nethermind committed:
fix: Set reduce_only field to true

ndatta-nethermind committed:
fix: Set reduce_only field to true

tyleroooo committed:
fix: use global node (#407)

'util' is unnecessary and breaks in some browser setups apparently

varex83 committed:
feat: remove some todos (#398)

ndatta-nethermind committed:
Validating transaction hash after placing order

ndatta-nethermind committed:
Validating transaction hash after placing order

ndatta-nethermind committed:
Validating transaction hash after placing order

ndatta-nethermind committed:
Validating transaction hash after placing order

ndatta-nethermind committed:
Iterating through all orders and checking with client id

ndatta-nethermind committed:
Iterating through all orders and checking with client id

ndatta-nethermind committed:
Iterating through all orders and checking with client id

ndatta-nethermind committed:
Iterating through all orders and checking with client id

ndatta-nethermind committed:
Added test case to check sell with no reduce by

ndatta-nethermind committed:
Added test case to check sell with no reduce by

ndatta-nethermind committed:
Added test case to check sell with no reduce by

ndatta-nethermind committed:
Added test case to check sell with no reduce by

ndatta-nethermind committed:
Fixed close position issue

ndatta-nethermind committed:
Fixed close position issue

ndatta-nethermind committed:
Fixed close position issue

ndatta-nethermind committed:
Fixed close position issue

therustmonk commented on v4-client-py-v2/dydx_v4_client/indexer/rest/constants.py:
```diff
@@ -5,8 +5,8 @@
 class OrderSide:
-    BUY = "BUY"
-    SELL = "SELL"
+    BUY = "SIDE_BUY"
```
I agree, I see in Rust it's still `BUY` and `SELL`. I think @varex83 didn't change them.

varex83 replied:
I remember we had some issues related to that, so it was the reason for change. Just checked in code, it doesn't affect a lot of things, since we translate it to the proto version of the OrderSide. Feel free to change it back

ndatta-nethermind replied:
Reverted back. Please check.

ndatta-nethermind committed:
Reverted order side changes

ndatta-nethermind committed:
Reverted order side changes

ndatta-nethermind committed:
Reverted order side changes

ndatta-nethermind committed:
Reverted order side changes

ndatta-nethermind committed:
Using order side from proto

ndatta-nethermind committed:
Using order side from proto

ndatta-nethermind committed:
Using order side from proto

ndatta-nethermind committed:
Using order side from proto

ndatta-nethermind committed:
Reverted changes in test_order

ndatta-nethermind committed:
Reverted changes in test_order

ndatta-nethermind committed:
Reverted changes in test_order

ndatta-nethermind committed:
Reverted changes in test_order

ndatta-nethermind committed:
Added builder code param and twap param in market order (#414)

alexey-kichin committed:
feat: add funding account balance checker to CI/CD pipeline (#412)

This commit introduces a new script, `balance_checker.py`, which checks
if the funding account has a sufficient balance for testing. The script
is integrated into the GitHub Actions workflow, ensuring that the
balance is verified during the CI/CD process. The minimum required
balance is set to 21 USDC.

alexey-kichin committed:
fix order tests (#415)

varex83 committed:
refactor: unify some method names with typescript client (#401)

gaonip commented:
Maybe in the future we can refactor the mutatiing node client test file, and split up instead of adding everything to this one file, but for we can merge

ndatta-nethermind committed:
Resolved merge conflicts

ndatta-nethermind committed:
Resolved merge conflicts

ndatta-nethermind committed:
Resolved merge conflicts

ndatta-nethermind committed:
Linted the changes

ndatta-nethermind committed:
Linted the changes

ndatta-nethermind committed:
Linted noble_client

jusbar23 committed:
feat: Adding missing order fields to protos (#417)

* Add ORRS to order proto and make clients consistent.

jusbar23 committed:
feat: Wire order router rev share in other clients (#418)

alexey-kichin committed:
Fix missing import (#419)

ndatta-nethermind committed:
Linted the changes
